src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -615,25 +668,33 @@ public bool ContainsKey(string key)
                 throw new ArgumentNullException(nameof(key));
             }
 
-            return _innerDictionary.ContainsKey(key);
+            return GetModelStateForKey(key)?.Visible ?? false;
         }
 
         /// <inheritdoc />
-        public void CopyTo(KeyValuePair<string, ModelStateEntry>[] array, int arrayIndex)
+        void ICollection<KeyValuePair<string, ModelStateEntry>>.CopyTo(
+            KeyValuePair<string, ModelStateEntry>[] array,
+            int arrayIndex)
         {
             if (array == null)
             {
                 throw new ArgumentNullException(nameof(array));
             }
 
-            ((ICollection<KeyValuePair<string, ModelStateEntry>>)_innerDictionary).CopyTo(array, arrayIndex);
+            if (arrayIndex < 0 || arrayIndex + Count > array.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(arrayIndex));
+            }
+
+            foreach (var item in this)
```

Comment:
Could do this recursively to avoid the enumerator allocation. Not sure how much we care about `CopyTo`.